examples/chess_position.py

```diff
@@ -0,0 +1,159 @@
+    (((board[1,4] == p) & (board[1,6] == p) & (board[1,7] == p) & (board[0,5] == b)).implies(Xor([board[0,6] == r,board[0,7] == r]))),
+    # If both bishops are stuck and pawns are in starting position, king and queen are also in starting position
+    (((board[1,1] == p) & (board[1,2] == p) & (board[1,3] == p) & (board[1,4] == p) & (board[1,5] == p) & (board[1,6] == p)).implies((board[0,3] == q) | (board[0,4] == k))),
+    (((board[6,1] == P) & (board[6,2] == P) & (board[6,3] == P) & (board[6,4] == P) & (board[6,5] == P) & (board[6,6] == P)).implies((board[7,3] == Q) | (board[7,4] == K))),
+    # If the pawns in column 2 or 7 didn't move, bishop of own color can't be in the corners of the board
+    (board[6,1] == P).implies(~((board[7,0] == B))),
+    (board[6,6] == P).implies(~((board[7,7] == B))),
+    (board[1,6] == p).implies(~((board[0,7] == b))),
+    (board[1,1] == p).implies(~((board[0,0] == b))),
+    # If the pawns in column 2 or 7 didn't move, bishop of other color can't be in the corners of the board excpet when promoted
+    ((board[6,1] == P) & (board[7,0] == B)).implies(sum(board[l,c] == p for l in range(0,8) for c in range(0,8)) <= 7),
+    ((board[6,6] == P) & (board[7,7] == B)).implies(sum(board[l,c] == p for l in range(0,8) for c in range(0,8)) <= 7),
+    ((board[1,6] == p) & (board[0,7] == b)).implies(sum(board[l,c] == P for l in range(0,8) for c in range(0,8)) <= 7),
+    ((board[1,1] == p) & (board[0,0] == b)).implies(sum(board[l,c] == P for l in range(0,8) for c in range(0,8)) <= 7),
+)
+
+# Put the index of the maximum value into the second row of the pieces variable
+for i in range(len(pieces_values)):
+    model += (((pieces[0,i] == pieces_values.item((i,0))) & (pieces[1,i] == 0)) | ((pieces[0,i] == pieces_values.item((i,1))) & (pieces[1,i] == 1)) | ((pieces[0,i] == pieces_values.item((i,2))) & (pieces[1,i] == 2)) | ((pieces[0,i] == pieces_values.item((i,3))) & (pieces[1,i] == 3)) | ((pieces[0,i] == pieces_values.item((i,4))) & (pieces[1,i] == 4)) | ((pieces[0,i] == pieces_values.item((i,5))) & (pieces[1,i] == 5)) \
+        | ((pieces[0,i] == pieces_values.item((i,6))) & (pieces[1,i] == 6)) | ((pieces[0,i] == pieces_values.item((i,7))) & (pieces[1,i] == 7)) | ((pieces[0,i] == pieces_values.item((i,8))) & (pieces[1,i] == 8)) | ((pieces[0,i] == pieces_values.item((i,9))) & (pieces[1,i] == 9)) | ((pieces[0,i] == pieces_values.item((i,10))) & (pieces[1,i] == 10)) | ((pieces[0,i] == pieces_values.item((i,11))) & (pieces[1,i] == 11)))
+
+# Put the pieces into the right place of the board
+index = 0
+for i in range(len(occupancy_in_board)):
+    if occupancy_in_board[i]:
+        model += (board[int(i/8), i%8] == pieces[1, index])
+        index += 1
+    else:
+        model += (board[int(i/8), i%8] == E)
+
+# Maximize the probabilites of the pieces to get the most likely board
+model.maximize(sum(pieces[0]))
```

Comment:
Based also on my comments before, regarding the variables for pieces that are not needed, I suggest to replace with

obj = sum(lp[v] for lp, v in zip(pieces_values, board[occupancy_in_board]))
model.maximize(obj)

This way, you directly linked the probabilities in pieces_values, with the cells in the board that need to be predicted.

Already tried the above and yield to the same solution.

And importantly, from 264 seconds that were needed to solve, it dropped to less than 0.5s.